usekase-coroutines/src/main/kotlin/guru/stefma/cleancomponents/usecase/coroutines/CoroutineUseCase.kt

```diff
@@ -1,17 +1,40 @@
 package guru.stefma.cleancomponents.usecase.coroutines
 
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.launch
+import kotlin.coroutines.CoroutineContext
+
 /**
- * A UseCase is a "actor" or "transformer" in the clean architecture.
+ * A UseCase is an "actor" or "transformer" in the Clean Architecture.
  *
- * Use this interface as a default implementation of an interface which have [params][P] and a [result][R].
+ * Use this base class to create a use case with input [parameters][Params] and an expected [result][Type].
  *
- * Simply call [execute] with the given [params][P] to get the [result][R].
+ * By default, once your overridden [execute] method returns, the callback lambdas passed to [invoke] will execute
+ * on the [main thread][Dispatchers.Main]. This requires a runtime dependency of coroutines that adds a main
+ * dispatcher. You can override [callbackDispatcher] to change this behavior.
  *
- * @param P the params you want to put in to the UseCase
- * @param R the result value which will be emitted by this UseCase
+ * @param Type The type which you expect to be returned by the use case.
+ * @param Params The parameters which the use case needs as an input.
  */
-interface CoroutineUseCase<R, P> {
+abstract class CoroutineUseCase<out Type, in Params> where Type : Any {
+
+    open val callbackDispatcher: CoroutineContext = Dispatchers.Main
 
-    suspend fun execute(params: P): R
+    /**
+     * Runs the actual logic of the use case.
+     */
+    abstract suspend fun execute(params: Params): Type
 
-}
+    suspend operator fun invoke(params: Params, onSuccess: (Type) -> Unit, onFailure: (Throwable) -> Unit) {
+        val result = runCatching { execute(params) }
```

Comment:
Catching exceptions inside use case will hide your own errors from you. Your app will not crash when you make an error inside the use case code and you are more likely to overlook it.

Also it mixes "domain errors" (like HTTP 404) and your own programmer mistakes in one pot.

Throwing exceptions for control flow is considered an anti-pattern/code smell:
https://softwareengineering.stackexchange.com/questions/189222/are-exceptions-as-control-flow-considered-a-serious-antipattern-if-so-why

Reply:
Also I don't think we can force the user to execute the coroutine in a certain way, like using `runCatching`. We should keep this out of the library. What if the user wants to `runBlocking` or `launch` the use case code?

Reply:
He could still call `execute` instead